android/app/src/main/java/org/uwaterloo/subletr/pages/profile/ProfilePageViewModel.kt

```diff
@@ -0,0 +1,217 @@
+		BehaviorSubject.createDefault(Optional.empty())
+	private val listingDetailsStream: BehaviorSubject<Optional<ListingDetails>> =
+		BehaviorSubject.createDefault(Optional.empty())
+	private val listingImageStream: BehaviorSubject<Optional<String>> =
+		BehaviorSubject.createDefault(Optional.empty())
+	val ratingStream: BehaviorSubject<Float> = BehaviorSubject.createDefault(0.0f)
+	val usersRatingStream: BehaviorSubject<Int> = BehaviorSubject.createDefault(0)
+
+	private val userDetailsStream: Observable<Result<GetUserResponse>> = userIdStream.map {
+		runBlocking {
+			runCatching {
+				userApi.userGet(it)
+			}.onSuccess { userResponse ->
+				nameStream.onNext("${userResponse.firstName} ${userResponse.lastName}")
+				ratingStream.onNext(userResponse.rating)
+				if (userResponse.listingId != null) {
+					listingIdStream.onNext(Optional.of(userResponse.listingId))
+					runCatching {
+						listingsApi.listingsDetails(
+							listingId = userResponse.listingId,
+							longitude = UWATERLOO_LONGITUDE,
+							latitude = UWATERLOO_LATITUDE,
+						)
+					}
+						.onSuccess { listing ->
+							listingDetailsStream.onNext(Optional.of(listing.details))
+							if (listing.details.imgIds.isNotEmpty()) {
+								runCatching {
+									listingsApi.listingsImagesGet(
+										listing.details.imgIds.first(),
+									)
+								}
+									.onSuccess { imageResponse ->
+										listingImageStream.onNext(Optional.of(imageResponse))
+									}
+									.onFailure {
+										Log.d("API ERROR", "Failed to get user's listing's image")
+									}
+							}
+						}
+						.onFailure {
+							Log.d("API ERROR", "Failed to get user's listing details")
+						}
+				}
+			}
+		}
+			.onFailure {
+				Log.d("API ERROR", "Failed to get user details")
+				navigationService.navHostController.popBackStack()
+			}
+	}
+		.onErrorResumeWith(Observable.never())
+		.subscribeOn(Schedulers.io())
+
+	private val listingImageBitmapStream: Observable<Optional<Bitmap>> = listingImageStream.map {
+		val nullableVersion = it.getOrNull()
+		if (nullableVersion != null) {
+			Optional.of(nullableVersion.base64ToBitmap())
+		}
+		else {
+			Optional.empty<Bitmap>()
+		}
+	}
+		.observeOn(Schedulers.computation())
+
+	private val avatarStream: Observable<Optional<String>> = userIdStream.map {
+		var avatar = Optional.empty<String>()
+		runCatching {
+			runBlocking {
+				userApi.userAvatarGet(it)
+			}
+		}
+			.onSuccess { img ->
+				avatar = Optional.of(img)
+			}
+			.onFailure {
+				avatar = Optional.empty<String>()
+			}
+		avatar
+	}
+		.onErrorResumeWith(Observable.never())
+		.subscribeOn(Schedulers.io())
+
+	private val avatarBitmapStream: Observable<Optional<Bitmap>> = avatarStream.map {
+		val nullableVersion = it.getOrNull()
+		if (nullableVersion != null) {
+			Optional.of(nullableVersion.base64ToBitmap())
+		}
+		else {
+			Optional.empty<Bitmap>()
+		}
+	}
+		.observeOn(Schedulers.computation())
```

Comment:
Same as above.

The `.observeOn(Schedulers.computation())` should be at the start of this stream and since we usually do `.io` stuff, you should probably switch it back to `.observeOn(Schedulers.io())` at the end, although in this case, it might be fine even if you don't do that last part.